### ISSUE TITLE
no-unused-var not work correctly with eval

### ISSUE THREAD

**Issue by jefer94** [eslint/eslint#9797]
**Tell us about your environment**

* **ESLint Version:**
4.14.0
* **Node Version:**
9.2.0
* **npm Version:**
5.5.1
* **Yarn Version:**
1.3.2

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
parserOptions:
  ecmaVersion: 8
  sourceType: "module"
  ecmaFeatures:
    impliedStrict: true
    jsx: true

globals: 
  Promise: true
  document: true
  Vector: true
  read: true

env:
  browser: true
  node: true
  es6: true

plugins: 
  - "promise"
  - "flowtype"
  - "babel"
  - "react"

extends: 
  - "plugin:flowtype/recommended"
  - "eslint:recommended"
  - "plugin:react/recommended"

parser: "babel-eslint"

rules:
  key-spacing: ["error", { "beforeColon": true, "afterColon": true, "align": { "beforeColon": true, "on": "colon" } }]
  no-multi-spaces: ["error", { exceptions: { "VariableDeclarator": true } }]
  react/jsx-uses-react: "error"
  react/jsx-uses-vars: "error"
  react/no-unescaped-entities: 0
  babel/new-cap: 1
  babel/object-curly-spacing: 0
  babel/no-invalid-this: 1
  babel/semi: 0
  no-cond-assign: 0
  strict: 0
  object-curly-spacing: ["error", "always"]
  array-bracket-spacing: ["error", "always"]
  indent: ["error", 2]
  quotes: ["error", "single", { "allowTemplateLiterals": true,  "avoidEscape": true }]
  no-unused-vars: ["error", { "vars": "all", "args": "none", "ignoreRestSiblings": false }]
  keyword-spacing: ["error", { before: true, after: true }]
  space-before-function-paren: ["error", "always"]
  eqeqeq: ["error", "always"]
  space-infix-ops: "error"
  comma-spacing: ["error", { "before": false, "after": true }]
  brace-style: ["error", "stroustrup"]
  curly: ["error", "multi"]
  handle-callback-err: "error"
  no-undef: "error"
  no-multiple-empty-lines: ["error", { "max": 1, "maxEOF": 1 }]
  operator-linebreak: ["error", "after"]
  block-spacing: "error"
  comma-dangle: ["error", "never"]
  comma-style: ["error", "last"]
  dot-location: ["error", "property"]
  eol-last: ["error", "always"]
  func-call-spacing: ["error", "never"]
  new-cap: ["error", { "newIsCap": true }]
  constructor-super: "error"
  no-caller: "error"
  no-class-assign: "error"
  no-constant-condition: "error"
  no-control-regex: "error"
  no-debugger: "error"
  no-delete-var: "error"
  no-dupe-args: "error"
  no-dupe-class-members: "error"
  no-dupe-keys: "error"
  no-duplicate-case: "error"
  no-duplicate-imports: "error"
  no-empty-character-class: "error"
  no-empty-pattern: "error"
  no-ex-assign: "error"
  no-extend-native: "error"
  no-extra-bind: "error"
  no-extra-boolean-cast: "error"
  no-extra-parens: "error"
  no-fallthrough: "error"
  no-floating-decimal: "error"
  no-func-assign: "error"
  no-global-assign: "error"
  no-implied-eval: "error"
  no-inner-declarations: "error"
  no-invalid-regexp: "error"
  no-irregular-whitespace: "error"
  no-iterator: "error"
  no-label-var: "error"
  no-labels: "error"
  no-lone-blocks: "error"
  no-mixed-spaces-and-tabs: "error"
  no-multi-str: "error"
  no-new: "error"
  no-new-func: "error"
  no-new-object: "error"
  no-new-require: "error"
  no-new-symbol: "error"
  no-new-wrappers: "error"
  no-obj-calls: "error"
  no-octal: "error"
  no-octal-escape: "error"
  no-path-concat: "error"
  no-proto: "error"
  no-redeclare: "error"
  no-regex-spaces: "error"
  no-self-assign: "error"
  no-self-compare: "error"
  no-sequences: "error"
  no-shadow-restricted-names: "error"
  no-sparse-arrays: "error"
  no-tabs: "error"
  no-template-curly-in-string: "error"
  no-this-before-super: "error"
  no-trailing-spaces: "error"
  no-undef-init: "error"
  no-unmodified-loop-condition: "error"
  no-unneeded-ternary: "error"
  no-unreachable: "error"
  no-unsafe-finally: "error"
  no-unsafe-negation: "error"
  no-useless-call: "error"
  no-useless-computed-key: "error"
  no-useless-constructor: "error"
  no-useless-escape: "error"
  no-with: "error"
  object-property-newline: "error"
  padded-blocks: ["error", "never"]
  rest-spread-spacing: ["error", "always"]
  semi-spacing: "error"
  no-useless-rename: "error"
  no-whitespace-before-property: "error"
  space-before-blocks: "error"
  space-in-parens: ["error", "never"]
  space-unary-ops: "error"
  spaced-comment: ["error", "always"]
  template-curly-spacing: "error"
  use-isnan: "error"
  valid-typeof: "error"
  wrap-iife: ["error", "outside"]
  yield-star-spacing: ["error", "both"]
  yoda: "error"
  semi: ["error", "never"]
  no-unexpected-multiline: "error"
  promise/no-return-wrap: "error"
  promise/param-names: "error"
  promise/no-native: "off"
  promise/no-nesting: "warn"
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
import Vector from './vector'
import transpiler from './transpiler'
import anythingInput from './anithing-input'

const read = (toRead) => {
  ...
}

const write = (input) => {
  ...
}

// result:
// write('hey applet!')
const exec = transpiler.parse(anythingInput)

try {
  eval(exec)
}
catch(e) {
  ...
}
```

```bash
yarn run test
```

**What did you expect to happen?**
I need a way to add a exception to no-unused-vars

**What actually happened? Please include the actual, raw output from ESLint.**
yarn run test
yarn run v1.3.2
$ eslint src/**/*.js main.js --fix

/home/jefer/prog/algorithm/src/components/console.js
   8:8  error  'Vector' is defined but never used         no-unused-vars
  14:9  error  'read' is assigned a value but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

**Comment by platinumazure** [comment]
This is a pretty unusual use case, so I don't know if we can make a change to this rule to support this case directly.

That said, you have a couple of options:

* Use [inline configuration comments](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments) to reconfigure or disable the rule in the file (including just on the lines with errors).
* Use the [`varsIgnorePattern`](https://eslint.org/docs/rules/no-unused-vars#varsignorepattern) rule option to specify what variables can be ignored, as a regex. (Combine this with the first option if you want to reconfigure the rule only for that file.)

**Comment by jefer94** [comment]
Thanks, The chosen solution due to the nature of the code was.

```js
/* eslint-disable no-unused-vars */
import Vector from './vector'
/* eslint-enable no-unused-vars */
import transpiler from './transpiler'
import anythingInput from './anithing-input'

/* eslint-disable no-unused-vars */
const read = (toRead) => {
  /* eslint-enable no-unused-vars */
  ...
}

/* eslint-disable no-unused-vars */
const write = (input) => {
  /* eslint-enable no-unused-var */
  ...
}

// result:
// write('hey applet!')
const exec = transpiler.parse(anythingInput)

try {
  eval(exec)
}
catch(e) {
  ...
}
```

**Comment by j-f1** [comment]
@jefer94 FYI you can write `// eslint-disable-next-line no-unused-vars` to have ESLint re-enable the rule for you after the next line.

**Comment by jefer94** [comment]
Thanks, a comment of single line look better.
